miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+        fn check_package(
+            // Path to Cargo.toml file of a package to check.
+            path: &str,
+            toolchain: &str,
+            extra_flags: &str,
+            all_targets: bool,
+            args: &[&str],
+        ) -> Result<()> {
+            let all_targets = all_targets.then_some("--all-targets");
+            cmd(
+                "cargo",
+                chain!(
+                    [&toolchain, "check", extra_flags, "--manifest-path", path,],
+                    all_targets,
+                    args.into_iter().cloned()
+                ),
+            )
+            .run()?;
+            Ok(())
+        }
+        // Check, and let caller control flags.
+        let toolchain = self.get_toolchain_override();
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let cargo_miri_manifest = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        let args = flags.iter().map(String::as_str).collect::<Vec<_>>();
+        check_package(
+            &miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            true,
+            &args,
+        )?;
+        check_package(
+            &cargo_miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            false,
+            &args,
+        )?;
+        Ok(())
+    }
+
+    /// Returns a toolchain override for use in command line (`+{toolchain_name}`).
+    fn get_toolchain_override(&self) -> String {
+        format!("+{}", &self.active_toolchain)
+    }
+    fn test(&mut self, flags: &[String]) -> Result<()> {
+        let bless = matches!(self.command, Subcommands::Bless { .. });
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let miri_manifest = miri_manifest.to_string_lossy().to_owned();
+        let args = flags.iter().map(String::as_str).collect::<Vec<_>>();
+        let toolchain = self.get_toolchain_override();
+        // First build and get a sysroot.
+        Self::build_package(&miri_manifest, &toolchain, &self.cargo_extra_flags, &[])?;
+        self.find_sysroot();
+        if bless {
+            self.extra_flags.insert("MIRI_BLESS".to_owned(), "Gesundheit".to_owned());
+        }
+        // Then test, and let caller control flags.
+        // Only in root project as `cargo-miri` has no tests.
+        cmd(
+            "cargo",
+            chain!(
+                [&toolchain, "test", &self.cargo_extra_flags, "--manifest-path", &miri_manifest],
+                args
+            ),
+        )
+        .run()?;
+        Ok(())
+    }
+
+    fn run_(&mut self, flags: &[String]) -> Result<()> {
+        use itertools::Itertools;
+        let target = flags.iter().tuple_windows().find(|(first, _)| first.as_str() == "--target");
+        if let Some(target) = target {
+            self.extra_flags.insert("MIRI_TEST_TARGET".to_owned(), target.1.clone());
```

Comment:
```suggestion
        if let Some((_, target)) = target {
            self.extra_flags.insert("MIRI_TEST_TARGET".to_owned(), target.clone());
```
Probably we want a `self.set_env` that takes `AsRef<OsStr>` to have less conversion noise everywhere